Source/Task/timer_cpp.cpp

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) Microsoft Corporation
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "pch.h"
+
+#include "timer.h"
+
+#include <algorithm>
+#include <chrono>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+using Deadline = std::chrono::high_resolution_clock::time_point;
+
+struct TimerEntry
+{
+    Deadline When;
+    PlatformTimer* Timer;
+
+    TimerEntry(Deadline d, PlatformTimer* t) : When{ d }, Timer{ t } {}
+};
+
+struct TimerEntryComparator
+{
+    bool operator()(TimerEntry const& l, TimerEntry const& r) noexcept
+    {
+        return l.When > r.When;
+    }
+};
+
+class TimerQueue
+{
+public:
+    bool LazyInit() noexcept;
+
+    void Add(PlatformTimer* timer, Deadline deadline) noexcept;
+    void Remove(PlatformTimer const* timer) noexcept;
+
+private:
+    void Worker() noexcept;
+
+    TimerEntry const& Peek() const noexcept;
+    TimerEntry Pop() noexcept;
+
+    std::mutex m_mutex;
+    std::condition_variable m_cv;
+    std::vector<TimerEntry> m_queue; // used as a heap
+    bool m_initialized = false;
+};
+
+namespace
+{
+
+std::once_flag g_timerQueueLazyInit;
+TimerQueue g_timerQueue;
+
+}
+
+bool TimerQueue::LazyInit() noexcept
+{
+    std::call_once(g_timerQueueLazyInit, [this]()
+    {
+        try
+        {
+            std::thread t([this]()
+            {
+                Worker();
+            });
+            t.detach();
+
+            m_initialized = true;
+        }
+        catch (...)
+        {
+            m_initialized = false;
+        }
+    });
+
+    return m_initialized;
+}
+
+void TimerQueue::Add(PlatformTimer* timer, Deadline deadline) noexcept
+{
+    {
+        std::lock_guard<std::mutex> lock{ m_mutex };
+
+        m_queue.emplace_back(deadline, timer);
+        std::push_heap(m_queue.begin(), m_queue.end(), TimerEntryComparator{});
+    }
+    m_cv.notify_all();
+}
+
+void TimerQueue::Remove(PlatformTimer const* timer) noexcept
+{
+    std::lock_guard<std::mutex> lock{ m_mutex };
+
+    for (auto& entry : m_queue)
+    {
+        if (entry.Timer == timer)
+        {
+            entry.Timer = nullptr;
```

Comment:
>entry.Timer = nullptr; [](start = 12, length = 22)

you don't want to also remove the entry?

Reply:
this is a heap, removing elements from it is non trivial.

---
In reply to: [192517405](https://github.com/Microsoft/libHttpClient/pull/181#discussion_r192517405) [](ancestors = 192517405)